pgxn/neon_rmgr/neon_rmgr.c

```diff
@@ -76,6 +80,9 @@ neon_rm_redo(XLogReaderState *record)
 		case XLOG_NEON_HEAP_MULTI_INSERT:
 			redo_neon_heap_multi_insert(record);
 			break;
+		case XLOG_NEON_LFC_PREWARM:
```

Comment:
> This PR implements prewarming of replica's LFC through normal replication protocol.
Primary periodically creates WAL records with state of LFC cache. By replaying this WAL records, replica can load this pages and so maintain the same state LFC as primary.

Do you have a full picture in mind of how the whole system will work?

I have doubts that this is the simplest and the most efficient one to go with because:

- Normally, I think it's reasonable to assume that primary and replica workloads are different. Well, that's 100% true because you do not have updates on replica, but even thinking about selects they might be very different. Look at how Neon uses Postgres internally -- Primary serves the main production operational workload, APIs, etc.; Replica is used for backoffice UIs, background jobs, and other analytical kind of things
- Yes, preferably, we do need to prewarm the replica and then do a switchover, but that's only needed shortly before the restart, while with this PR we will be writing a lot of WAL all the time. It's probably possible to overcome by something setting this GUC only before we want to do a restart and disabling it shortly after restart is finished
- Restart, or more precisely 'scheduled graceful restart', is not the only problem, there at least two others -- wake up after scale-to-zero and unexpected/unscheduled/unorchestrated restart (e.g. k8s node went down), in both cases we do want a prewarm as well

Saying that, I'm not sure that this approach will eventually help us with solving all the cache prewarming problems on compute. And maybe developing dump/restore API on compute + S3 persistence as we discussed is still the best way to go, then:
1. The case of the scheduled restart could be covered by something orchestrating it by calling dump on RO and restore on RW
2. While unexpected restart will be handled by compute self-serve. It will request the cache content from S3 lazily at start and do prewarming

In both cases, I assume that having slightly outdated caches is OK. While that's true for 2., i.e. compute will request the latest version of page from the storage, we just need to know that page N was in the cache before restart; in case 1. it might be not true, because if cache will contain some older version of the page while it's already updated on primary it might cause corruption-like problems after switchover, right? So we might still need something that you proposes here to handle the case 1.

What do you think?

Reply:
I mostly try to address case of restarting huge node with intensive workload. If node is small or almost idle, then there is no big need in prewarming. Or at least it can be do in other way.

So scenario I take in my head wasn't actually my scenario - it was mentioned by Star at the very beginning of this discussion:
- We have running node which we want (need) to restart
-  We spawn replica and prewarm it. There is no any workload on the replica. So it just prewarming. Once prewarming is completed (actually it is not quite easy to determine when it is necessary to stop prewarming if there is stream of permanent updates on primary), we stop primary and promote replica. Definitely we expect that world at new node will be the same as on old node, so we try to load it's LFC cache with the same content as old node.

What about other scenarios you mentioned? I doubt that huge node under intensive workload will be sometimes scaled to zero (terminated). But crash of the node is definitely possible. In this case we can do what vanilla Postgres does: permanently maintain hot-standby replica. The mechanism proposed in this PR allows to keep it in sync with primary. It is not intended to be used for scaling read-only workload. It is just standby making it possible to perform fast master replacement at each moment of time.

I am not 100% sure that it is the best way to perform prewarming. And it is not still clear to me how critical is prewarming for us (how long it will take for huge node to reach previous level of performance under high load after restart). If there are large number of quires to the database, then may be they can warm cache much more efficiently than pre-warming done by single background worker.

In any case, there seems to be many concerns against using LFC for prewarming and other similar stuff. This PR illustrates how it is possible without LFC and without any other changes in Postgres. Should we follow this way? I don't know. But I really first want to know results of tests.